src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,239 @@
+    end
+    :(nothing)
+end
+
+Base.@pure triangularnumber(N::Int) = div(N * (N + 1), 2)
+Base.@pure triangularroot(L::Int) = div(isqrt(8 * L + 1) - 1, 2) # from quadratic formula
+
+lowertriangletype(::Type{SSymmetricCompact{N, T, L}}) where {N, T, L} = SVector{L, T}
+lowertriangletype(::Type{SSymmetricCompact{N, T}}) where {N, T} = SVector{triangularnumber(N), T}
+lowertriangletype(::Type{SSymmetricCompact{N}}) where {N} = SVector{triangularnumber(N)}
+
+@inline (::Type{SSymmetricCompact{N, T}})(lowertriangle::SVector{L}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+@inline (::Type{SSymmetricCompact{N}})(lowertriangle::SVector{L, T}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+
+@inline function SSymmetricCompact(lowertriangle::SVector{L, T}) where {T, L}
+    N = triangularroot(L)
+    SSymmetricCompact{N, T, L}(lowertriangle)
+end
+
+@generated function (::Type{SSymmetricCompact{N, T, L}})(a::Tuple) where {N, T, L}
+    expr = Vector{Expr}(undef, L)
+    i = 0
+    for col = 1 : N, row = col : N
+        index = N * (col - 1) + row
+        expr[i += 1] = :(a[$index])
+    end
+    quote
+        @_inline_meta
+        @inbounds return SSymmetricCompact{N, T, L}(SVector{L, T}(tuple($(expr...))))
+    end
+end
+
+@inline function (::Type{SSymmetricCompact{N, T}})(a::Tuple) where {N, T}
+    L = triangularnumber(N)
+    SSymmetricCompact{N, T, L}(a)
+end
+
+@inline (::Type{SSymmetricCompact{N}})(a::Tuple) where {N} = SSymmetricCompact{N, promote_tuple_eltype(a)}(a)
+@inline (::Type{SSymmetricCompact{N}})(a::NTuple{M, T}) where {N, T, M} = SSymmetricCompact{N, T}(a)
+@inline SSymmetricCompact(a::StaticMatrix{N, N, T}) where {N, T} = SSymmetricCompact{N, T}(a)
+
+@inline (::Type{SSC})(a::SSymmetricCompact) where {SSC <: SSymmetricCompact} = SSC(a.lowertriangle)
+@inline (::Type{SSC})(a::SSC) where {SSC <: SSymmetricCompact} = a
+
+@inline (::Type{SSC})(a::AbstractVector) where {SSC <: SSymmetricCompact} = SSC(convert(lowertriangletype(SSC), a))
+
+# TODO: similar_type overload?
+
+@generated function _symmetric_compact_indices(::Val{N}) where N
+    # Returns a Tuple{Pair{Int, Bool}} I such that for linear index i,
+    # * I[i][1] is the index into the lowertriangle field of an SSymmetricCompact{N};
+    # * I[i][2] is true iff i is an index into the lower triangle of an N × N matrix.
+    indexmat = Matrix{Pair{Int, Bool}}(undef, N, N)
+    i = 0
+    for col = 1 : N, row = 1 : N
+        indexmat[row, col] = if row >= col
+            (i += 1) => true
+        else
+            indexmat[col, row][1] => false
+        end
+    end
+    quote
+        Base.@_inline_meta
+        return $(tuple(indexmat...))
+    end
+end
+
+Base.@propagate_inbounds function Base.getindex(a::SSymmetricCompact{N}, i::Int) where {N}
+    I = _symmetric_compact_indices(Val(N))
+    j, lower = I[i]
+    @inbounds x′ = a.lowertriangle[j]
```

Comment:
Precedent for this was

https://github.com/JuliaLang/julia/blob/5b637df34396034b0dd353e603ab3d61322369fb/base/broadcast.jl#L823

Suggestions for another name?

Reply:
Yeah, I know people sometimes use `\prime`. My point is only this made me have to stare and think about what `adjoint` was doing on the left hand side (if only for twenty seconds; it would have been longer the first time).

I think that's an indication we should use another name; `x` would be fine?

Reply:
I wanted symmetry with `setindex` where `x` is an argument. Changed to `value`.